server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -710,9 +715,13 @@ public void onNoLongerMaster() {
         if (!lifecycle.started()) {
             return;
         }
-        try {
+        final ThreadContext threadContext = threadPool.getThreadContext();
+        final Supplier<ThreadContext.StoredContext> supplier = threadContext.newRestorableContext(false);
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            threadContext.markAsSystemContext();
```

Comment:
instead of doing this, I wonder if we should invest in a SystemThreadContext that we give to the executor and it does is under the system context. Alternatively we can have a "no context" executor that simply doesn't set any context. @tvernum do you see any other usage for this?

Reply:
I'm not sure if I understand this correctly. Would you have different executors based on whether they switch to the system context or stay in default context? Looking at the other usages of `threadContext.markAsSystemContext();`, all of them seem to use executors that cannot be categorized as system or non-system (i.e. generic threadpool, ...).

Reply:
my suggestion referred to the entire pattern of generating a new context and then marking it as system and deal with on the executor level. This has two potential upsides:
1) You don't need to remember to do this everywhere you submit a task (not so important here as there's just one place).
2) We don't do all the "clean & capture and then restore in the thread pool" song and dance.

Since this is not a high throughput usage and we only have one place we submit tasks it doesn't have such a big impact on `MasterService` but might be useful for other things that use an internal thread pool executor.